### PR TITLE
Adds PlusOneBackupSelector to RAMCloud

### PR DESCRIPTION
**Pull request by OferMania**
This gives us deterministic selection for the backupServerId, independent of the replicationId.
We also add test_backup_server.py with Fault Tolerance tests for when a backup server goes down with PlusOneBackup enabled.

Resolves NF-351

### CODE REVIEW

File: patches/plus_one.patch [hunk 1]
```diff
@@ -0,0 +1,777 @@
...
+     : context(context)
+     , numReplicas(numReplicas)
+@@ -61,12 +66,17 @@ ReplicaManager::ReplicaManager(Context* context,
+     , failureMonitor(context, this)
+     , replicationCounter()
+     , useMinCopysets(useMinCopysets)
++    , usePlusOneBackup(usePlusOneBackup)
+     , allowLocalBackup(allowLocalBackup)
+ {
+     if (useMinCopysets) {
```
Comment (vexingcodes) [inline patches/plus_one.patch:558]: Probably good to do some config-checking here.

```
if (useMinCopysets && usePlusOneBackup) {
    throw std::invalid_argument(
        "Can only use one of min-copysets and plus-one backup strategies, but both were specified.");
}
```

File: patches/plus_one.patch [hunk 2]
```diff
@@ -0,0 +1,777 @@
...
++/**
++ * Selects backups to store replicas starting with (masterServerId+1)%n
++ */
++class PlusOneBackupSelector : public BackupSelector {
++  PUBLIC:
++    explicit PlusOneBackupSelector(Context* context,
++                                       const ServerId* serverId,
++                                       uint32_t numReplicas,
++                                       bool allowLocalBackup);
++    ServerId selectSecondary(uint32_t numBackups, const ServerId backupIds[]);
```
Comment (vexingcodes) [inline patches/plus_one.patch]: Add the `virtual` and/or `override` keywords to this function declaration.

File: patches/plus_one.patch [hunk 3]
```diff
@@ -0,0 +1,777 @@
...
++} // namespace RAMCloud
++
++#endif
+diff --git a/src/PlusOneBackupSelectorTest.cc b/src/PlusOneBackupSelectorTest.cc
+new file mode 100644
+index 00000000..ab0485ab
+--- /dev/null
++++ b/src/PlusOneBackupSelectorTest.cc
+@@ -0,0 +1,101 @@
++/* Copyright (c) 2009-2016 Stanford University
```
Comment (vexingcodes) [inline patches/plus_one.patch]: 2019 is the current year. I have no idea how copyright statements work though. I don't know if we should just omit this entirely on files that we create? :man_shrugging:

**Review by vexingcodes — CHANGES_REQUESTED** [review]
Looks good, just minor comments.

File: patches/plus_one.patch (hunk 2 above)
Comment (OferMania) [reply patches/plus_one.patch]: recommended c++11 convention is to use just `override`, so I'm sticking with that.

File: patches/plus_one.patch (hunk 3 above)
Comment (OferMania) [reply patches/plus_one.patch]: I'm not sure either, but will change date to 2019 for now. When I submit for RAMCloud code review, we can find out at that point. I doubt it matters for a patchset internal to us.

File: patches/plus_one.patch (hunk 1 above)
Comment (OferMania) [reply patches/plus_one.patch:558]: Will add

**Review by vexingcodes — APPROVED** [review]
Looks good!

**Pull request merged**